model/cells/zhong_cell.hpp

```diff
@@ -0,0 +1,199 @@
+
+            // A proportion of the previous day's infection recovers, leading to a new proportion
+            // of the population that is currently infected at the current day of infection
+            curr_inf *= 1 - recovery_rates[res.phase][i-1];
+
+            res.infected[i] = std::round(curr_inf * precDivider) / precDivider;
+
+            // The amount of susceptible does not include the amount the infected proportion of the population
+            new_s -= res.infected[i];
+        }
+
+        // The people on the first day of infection are equal to the number of infections from the susceptible population
+        res.infected[0] = new_i;
+
+        // The susceptible population does not include infected individuals
+        new_s -= new_i;
+
+        // Equation 6a
+        for (int i = res.num_rec - 1; i > 0; --i)
+        {
+            // Each day of the recovered is the value of the previous day. The population on the last day is
+            // now susceptible; this is implicitly done already as the susceptible value was set to 1.0 and the
+            // population on the last day of recovery is never subtracted from the susceptible value.
+            res.recovered[i] = res.recovered[i-1];
+            new_s -= res.recovered[i];
+        }
+
+        // The people on the first day of recovery are those that were on the last stage of infection
+        // in the previous time step.
+        res.recovered[0] = new_r;
+
+        // The susceptible population does not include the recovered population
+        new_s -= new_r;
+
+        if(new_s > -0.001 && new_s < 0) new_s = 0;  // float precision issues
+
+        assert(new_s >= 0);
+        res.susceptible = new_s;
+
+        res.phase = next_phase(res.phase);
+
+        return res;
+    }
+    // It returns the delay to communicate cell's new state.
+    T output_delay(sir const &cell_state) const override {
+        return 1;
+    }
+
+    void check_valid_vicinity() {
+
+        sir cstate = state.current_state;
+
+        for(const auto& neighbor : neighbors) {
+
+            sir nstate = state.neighbors_state.at(neighbor);
+            vicinity v = state.neighbors_vicinity.at(neighbor);
+
+            assert(v.correlation >= 0 && v.correlation <= 1);
+        }
+    }
+
+    void check_valid_state() const {
+        sir init = state.current_state;
+        float sum = init.susceptible;
+        assert(init.susceptible >= 0 && init.susceptible <= 1);
+        for(int i=0; i < init.num_inf; i++) {
+            assert(init.infected[i] >= 0 && init.infected[i] <= 1);
+            sum += init.infected[i];
+        }
+        for(int i=0; i < init.num_rec; i++) {
+            assert(init.recovered[i] >= 0 && init.recovered[i] <= 1);
+            sum += init.recovered[i];
+        }
+        assert(sum == 1);
+    }
+
+    float new_infections() const {
+        float inf = 0;
+        sir cstate = state.current_state;
+
+        // internal infected
```

Comment:
Look here: you are using the cell's current state and preprocessing the next state using just your state...